File: src/java_version/arrays/MergeSort.java
package java_version.arrays;

public class MergeSort {

    public static void main(String[] args) {
        int[] array = Utils.arrayOfSevenElements();
        long startTime = System.nanoTime();
        mergeSort(array, 0, array.length);
        System.out.print("Sorted: " + (System.nanoTime() - startTime) + "\n");
        for (int i : array) {
            System.out.print("" + i + "\n");
        }
    }

    static void mergeSort(int[] array, int start, int end) {
        if (end - start < 2) {
            return;
        }

        int mid = (start + end) / 2;

        mergeSort(array, start, mid);
        mergeSort(array, mid, end);
        mergeArrayS(array, start, end, mid);
    }

    static void mergeArrayS(int[] array, int start, int end, int mid) {
        if (array[mid - 1] <= array[mid]) {
            return;
        }

        int i = start;
        int j = mid;
        int tempIndex = 0;
        int[] tempArray = new int[end - start];

        while (i < mid && j < end) {
            tempArray[tempIndex++] = array[i] <= array[j] ? array[i++] : array[j++];
        }

        System.arraycopy(array, i, array, start + tempIndex, mid - i);
        System.arraycopy(tempArray, 0, array, start, tempIndex);
    }

}

File: src/kotlin_version/arrays/MergeSort.kt
package kotlin_version.arrays

import java_version.arrays.Utils

val arrayOfIntegers: IntArray = Utils.arrayOfMillions()

fun main() {
    val startTime = System.nanoTime()

    mergeSort(arrayOfIntegers, 0, arrayOfIntegers.size)

    println("Sorted: ${System.nanoTime() - startTime}")
}

fun mergeSort(array: IntArray, start: Int, end: Int) {
    if (end - start < 2) {
        return
    }

    val mid = (start + end) / 2

    mergeSort(array, start, mid)
    mergeSort(array, mid, end)
    merge(array, start, mid, end)
}

fun merge(array: IntArray, start: Int, mid: Int, end: Int) {
    if (array[mid - 1] < array[mid]) {
        return
    }

    var i = start
    var j = mid
    var tempIndex = 0
    val tempArray = IntArray(end - start)

    while (i < mid && j < end) {
        tempArray[tempIndex++] = if (arrayOfIntegers[i] < arrayOfIntegers[j]) arrayOfIntegers[i++] else arrayOfIntegers[j++]
    }

    System.arraycopy(array, i, array, start + tempIndex, mid - i)
    System.arraycopy(tempArray, 0, array, start, tempIndex)
}